google/resource_composer_environment.go

```diff
@@ -0,0 +1,974 @@
+	}
+
+	log.Printf("[DEBUG] Finished creating Environment %q: %#v", d.Id(), op)
+
+	if err := resourceComposerEnvironmentPostCreateUpdate(updateOnlyEnv, d, config); err != nil {
+		return err
+	}
+
+	return resourceComposerEnvironmentRead(d, meta)
+}
+
+func resourceComposerEnvironmentRead(d *schema.ResourceData, meta interface{}) error {
+	config := meta.(*Config)
+
+	envName, err := resourceComposerEnvironmentName(d, config)
+	if err != nil {
+		return err
+	}
+
+	res, err := config.clientComposer.Projects.Locations.Environments.Get(envName.resourceName()).Do()
+	if err != nil {
+		return handleNotFoundError(err, d, fmt.Sprintf("ComposerEnvironment %q", d.Id()))
+	}
+
+	// Set from d.GetProject
+	if err := d.Set("project", envName.Project); err != nil {
+		return fmt.Errorf("Error reading Environment: %s", err)
+	}
+	// Set from d.GetRegion
+	if err := d.Set("region", envName.Region); err != nil {
+		return fmt.Errorf("Error reading Environment: %s", err)
+	}
+	if err := d.Set("name", GetResourceNameFromSelfLink(res.Name)); err != nil {
+		return fmt.Errorf("Error reading Environment: %s", err)
+	}
+	if err := d.Set("config", flattenComposerEnvironmentConfig(res.Config)); err != nil {
+		return fmt.Errorf("Error reading Environment: %s", err)
+	}
+	if err := d.Set("uuid", res.Uuid); err != nil {
+		return fmt.Errorf("Error reading Environment: %s", err)
+	}
+	if err := d.Set("state", res.State); err != nil {
+		return fmt.Errorf("Error reading Environment: %s", err)
+	}
+	if err := d.Set("create_time", res.CreateTime); err != nil {
+		return fmt.Errorf("Error reading Environment: %s", err)
+	}
+	if err := d.Set("update_time", res.UpdateTime); err != nil {
+		return fmt.Errorf("Error reading Environment: %s", err)
+	}
+	if err := d.Set("labels", res.Labels); err != nil {
+		return fmt.Errorf("Error reading Environment: %s", err)
+	}
+	return nil
+}
+
+func resourceComposerEnvironmentUpdate(d *schema.ResourceData, meta interface{}) error {
+	tfConfig := meta.(*Config)
+
+	d.Partial(true)
+	if d.HasChange("config") {
+		config, err := expandComposerEnvironmentConfig(d.Get("config"), d, tfConfig)
+		if err != nil {
+			return err
+		}
+
+		if d.HasChange("config.0.software_config.0.airflow_config_overrides") {
+			patchObj := &composer.Environment{
+				Config: &composer.EnvironmentConfig{
+					SoftwareConfig: &composer.SoftwareConfig{
+						AirflowConfigOverrides: make(map[string]string),
+					},
+				},
+			}
+			if config != nil && config.SoftwareConfig != nil && len(config.SoftwareConfig.AirflowConfigOverrides) > 0 {
+				patchObj.Config.SoftwareConfig.AirflowConfigOverrides = config.SoftwareConfig.AirflowConfigOverrides
+			}
+
+			err = resourceComposerEnvironmentPatchField("config.softwareConfig.airflowConfigOverrides", patchObj, d, tfConfig)
+			if err != nil {
+				return err
+			}
+			d.SetPartial("config")
+		}
+
+		if d.HasChange("config.0.software_config.0.env_variables") {
+			patchObj := &composer.Environment{
```

Comment:
TBH I'm not sure - I am using expandComposerEnvironmentConfig above so there is a little reuse?  I didn't want to deal with the case that they might not enjoy a particular field being included somewhere, so I thought it would be overall less to worry about if I used the expander and then created "immutable" throwaway object with only the fields being updated, in case.